opsdroid/connector/matrix/connector.py

```diff
@@ -223,20 +246,35 @@ async def get_nick(self, roomid, mxid):
         setting.
         """
         if self.room_specific_nicks:
-            try:
-                return await self.connection.get_room_displayname(roomid, mxid)
-            except Exception:  # pylint: disable=W0703
-                # Fallback to the non-room specific one
-                logging.exception("Failed to lookup room specific nick for %s.", mxid)
+            if roomid is None:
+                return mxid
+
+            res = await self.connection.joined_members(roomid)
+            if isinstance(res, nio.JoinedMembersError):
+                logging.exception("Failed to lookup room members for %s.", roomid)
+                # fallback to profile
+                res = await self.connection.get_displayname(mxid)
+                if isinstance(res, nio.ProfileGetDisplayNameError):
+                    _LOGGER.error("Failed to lookup nick for %s.", mxid)
+                    return mxid
+                if res.displayname is None:
+                    return mxid
+                return res.displayname
+
+            for member in res.members:
+                if member.user_id == mxid:
+                    return member.display_name
 
-        try:
-            return await self.connection.get_display_name(mxid)
-        except MatrixRequestError as mre:
-            # Log the error if it's not the 404 from the user not having a nick
-            if mre.code != 404:
-                logging.exception("Failed to lookup nick for %s.", mxid)
             return mxid
 
+        res = await self.connection.get_displayname(mxid)
+        if isinstance(res, nio.ProfileGetDisplayNameError):
+            _LOGGER.error("Failed to lookup nick for %s.", mxid)
+            return mxid
```

Comment:
Since this "if error: log error, return" structure is coming up a lot here, I wonder if it's worth defining a function or a context manager to wrap it, just to make the code in the connector a little tidier? Probably not worth worrying about for this PR but something to consider going forward.

Reply:
hmm not sure how to write this as every function has its own error class that is returned